stdlib/public/Atomics/UnsafeAtomic.swift.gyb

```diff
@@ -0,0 +1,376 @@
+  /// applying the specified memory ordering.
+  ///
+  /// This operation performs the following algorithm as a single atomic
+  /// transaction:
+  ///
+  /// ```
+  /// atomic(self) { currentValue in
+  ///   let original = currentValue
+  ///   guard original == expected else { return (false, original) }
+  ///   currentValue = desired
+  ///   return (true, original)
+  /// }
+  /// ```
+  ///
+  /// This method implements a "strong" compare and exchange operation
+  /// that does not permit spurious failures.
+  ///
+  /// - Parameter expected: The expected current value.
+  /// - Parameter desired: The desired new value.
+  /// - Parameter ordering: The memory ordering to apply on this operation.
+  /// - Returns: A tuple `(exchanged, original)`, where `exchanged` is true if
+  ///   the exchange was successful, and `original` is the original value.
+  @_semantics("atomics.requires_constant_orderings")
+  @_transparent @_alwaysEmitIntoClient
+  public func compareExchange(
+    expected: Value,
+    desired: __owned Value,
+    ordering: AtomicUpdateOrdering
+  ) -> (exchanged: Bool, original: Value) {
+    let (exchanged, original) = Value._AtomicStorage._atomicCompareExchange(
+      expected: Value._encodeAtomicStorage(for: expected),
+      desired: Value._encodeAtomicStorage(for: desired),
+      at: _ptr,
+      ordering: ordering)
+    return (exchanged, Value._decodeAtomicStorage(original))
+  }
+
+  /// Perform an atomic compare and exchange operation on the current value,
+  /// applying the specified success/failure memory orderings.
+  ///
+  /// This operation performs the following algorithm as a single atomic
+  /// transaction:
+  ///
+  /// ```
+  /// atomic(self) { currentValue in
+  ///   let original = currentValue
+  ///   guard original == expected else { return (false, original) }
+  ///   currentValue = desired
+  ///   return (true, original)
+  /// }
+  /// ```
+  ///
+  /// The `ordering` argument specifies the memory ordering to use when the
+  /// operation manages to update the current value, while `failureOrdering`
+  /// will be used when the operation leaves the value intact.
+  ///
+  /// This method implements a "strong" compare and exchange operation
+  /// that does not permit spurious failures.
+  ///
+  /// - Parameter expected: The expected current value.
+  /// - Parameter desired: The desired new value.
+  /// - Parameter ordering: The memory ordering to apply on this operation.
+  /// - Returns: A tuple `(exchanged, original)`, where `exchanged` is true if
+  ///   the exchange was successful, and `original` is the original value.
+  @_semantics("atomics.requires_constant_orderings")
+  @_transparent @_alwaysEmitIntoClient
+  public func compareExchange(
+    expected: Value,
+    desired: __owned Value,
+    ordering: AtomicUpdateOrdering,
+    failureOrdering: AtomicLoadOrdering
+  ) -> (exchanged: Bool, original: Value) {
```

Comment:
In these APIs, I find some of the names confusing.

* `expected` and `original` ➡️ `oldValue`
* `desired` ➡️ `newValue`
* `exchanged` ➡️ `success`
* `ordering` ➡️ `successOrdering` (iff the API has `failureOrdering`)
* possibly an `inout` parameter?

```suggestion
  public func compareExchange(
    oldValue: inout Value,
    newValue: __owned Value,
    successOrdering: AtomicUpdateOrdering,
    failureOrdering: AtomicLoadOrdering
  ) -> Bool {
```

Reply:
I like `successOrdering` -- I'm pushing a commit that switches to that.

Having an `inout` parameter was deliberately rejected, based on negative experience with the corresponding C++ construct.

I don't see how `expected`/`desired` would be confusing, and they match preexisting terminology. But this may be a topic for the evolution thread.